motr/client.c

```diff
@@ -830,84 +831,116 @@ M0_INTERNAL int m0_op_init(struct m0_op *op,
+					"bvec->ov_buf[%d]=%p, bvec->ov_vec.v_count[%u]=%u",
+					curr_context, i, bvec->ov_buf[i], i,
+					(uint32_t)bvec->ov_vec.v_count[i]);
+		}
 	}
 
-	/* if caller want checksum without seed and with seed, in this case
-	 * 'pi_value_without_seed' will be used to send wihout seed checksum.
+	/* If caller wants checksum without seed and with seed, caller needs to
+	 * pass 'pi_value_without_seed'. pi_value will be used to return with
+	 * seed checksum. 'pi_value_without_seed' will be used to return non
+	 * seeded checksum.
 	 */
-	if (pi_value_without_seed && (flag & M0_PI_CALC_FINAL)) {
-		unsigned char digest[sizeof(MD5_CTX)];
-		memcpy(&digest, curr_digest, sizeof(MD5_CTX));
-
+	if (pi_value_without_seed != NULL) {
 		/* 
-		 * NOTE: MD5_final() changes the digest itself and curr_digest
+		 * NOTE: MD5_final() changes the context itself and curr_context
 		 * should not be finalised, thus copy it and use it for MD5_final
 		 */
-
-		MD5_Final(pi_value_without_seed, digest);
+		memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
+
+		rc = MD5_Final(pi_value_without_seed, &context);
+		if (rc != 1) {
+			return M0_ERR_INFO(rc, "MD5_Final failed"
+					"pi_value_without_seed=%p"
+					"curr_context=%p",
+					pi_value_without_seed, curr_context);
+		}
 	}
 
-	/* if seed is passed, memcpy and update the digest calculated so far.
+	/* if seed is passed, memcpy and update the context calculated so far.
 	 * calculate checksum with seed, set the pi_value with seeded checksum.
-	 * NOTE: curr_digest will always have digest without seed.
+	 * If seed is not passed than memcpy context and calculate checksum
+	 * without seed, set the pi_value with unseeded checksum.
+	 * NOTE: curr_context will always have context without seed.
 	 */
-	if (seed) {
-		unsigned char seed_digest[sizeof(MD5_CTX)];
-		memcpy(&seed_digest, curr_digest, sizeof(MD5_CTX));
+	memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
+
+	if (seed != NULL) {
 
-		MD5_Update(&seed_digest, &seed, sizeof(struct m0_pi_seed));
-		MD5_Final(pi->pi_value, &seed_digest);
+		/*
+		 * seed_str have string represention for 3 uint64_t(8 bytes)
+		 * range for uint64_t is 0 to 18,446,744,073,709,551,615 at
+		 * max 20 chars per var, for three var it will be 3*20, +1 '\0'.
+		 * seed_str needs to be 61 bytes, round off and taking 64 bytes.
+		 */
+		char seed_str[64] = {'\0'};
+		snprintf(seed_str, sizeof(seed_str), "%"PRIx64"%"PRIx64"%"PRIx64,
+				seed->obj_id.f_container, seed->obj_id.f_key,
+				seed->data_unit_offset);
+		rc = MD5_Update(&context, (unsigned char *)seed_str,
+				sizeof(seed_str));
+		if (rc != 1) {
+
+			return M0_ERR_INFO(rc, "MD5_Update fail curr_context=%p"
+					"f_container %"PRIx64"f_key %"PRIx64
+					"data_unit_offset %"PRIx64"seed_str %s",
+					curr_context, seed->obj_id.f_container,
+					seed->obj_id.f_key,
+					seed->data_unit_offset,
+					(char *)seed_str);
+		}
 	}
 
-#endif
+	if (!(flag & M0_PI_SKIP_CALC_FINAL)) {
```

Comment:
If we are passing the seed, we will always intend for seeded_final_checksum

Reply:
By default if this flag is not set this it will compute final checksum, this is mainly for optimization.